### PR TITLE
perf: improve performance of IndicesBuilder compute ivf partitions

### PR DESCRIPTION
**Pull request by eddyxu**

### CODE REVIEW

File: python/python/lance/torch/data.py [hunk 1]
```diff
@@ -13,6 +13,7 @@
 from typing import Dict, Iterable, List, Literal, Optional, Union
 
 import pyarrow as pa
+from line_profiler import profile
```
Comment (chebbyChefNEQ) [inline python/python/lance/torch/data.py:16]: remove?

File: python/python/lance/torch/data.py [hunk 2]
```diff
@@ -52,17 +51,20 @@ def _to_tensor(
             pa.types.is_floating(arr.type.value_type)
             or pa.types.is_integer(arr.type.value_type)
         ):
-            np_arrs = arr.to_numpy(zero_copy_only=False)
-            np_tensor = np.stack(np_arrs)
-            del np_arrs
-            tensor = torch.tensor(np_tensor)
+            np_tensor = arr.values.to_numpy(zero_copy_only=True).reshape(
```
Comment (chebbyChefNEQ) [inline python/python/lance/torch/data.py:54]: is non chunked array garueenteed here?

File: python/python/lance/torch/data.py [hunk 3]
```diff
@@ -216,31 +223,16 @@ def __init__(
...
-                else:
-                    sampler = FullScanSampler()
-            elif shard_granularity == "batch":
-                sampler = ShardedBatchSampler(rank, world_size)
-            elif shard_granularity == "fragment":
-                sampler = ShardedFragmentSampler(rank, world_size)
-            else:
-                raise ValueError("Invalid shard_granularity: {}")
+        if rank is not None and world_size is not None:
+            warnings.warn("rank and world_size are deprecated")
```
Comment (chebbyChefNEQ) [inline python/python/lance/torch/data.py:230]: nit: specify the warning type as https://docs.python.org/3/library/exceptions.html#DeprecationWarning

File: python/python/lance/vector.py [hunk 4]
```diff
@@ -209,6 +213,8 @@ def compute_partitions(
     dst_dataset_uri: Union[str, Path], optional
         The path to store the partitions.  If not specified a random
         directory is used instead
+    num_workers: int, default 4
+        Number of workers for PyTorch DataLoader.
```
Comment (chebbyChefNEQ) [inline python/python/lance/vector.py:217]: there is no arg called `num_workers`

File: python/python/lance/torch/data.py (hunk 2 above)
Comment (eddyxu) [reply python/python/lance/torch/data.py:54]: Our lance rust core does not really have concept of `chunked array`. This is read from `lance.dataset`, i think we wont have chunked table in this case?

Comment (chebbyChefNEQ) [reply python/python/lance/torch/data.py:54]: I'm pretty sure I have see chunked arraies being returned. Let me test

File: python/python/lance/torch/data.py (hunk 1 above)
Comment (eddyxu) [reply python/python/lance/torch/data.py:16]: done

File: python/python/lance/vector.py (hunk 4 above)
Comment (eddyxu) [reply python/python/lance/vector.py]: Ok, this is left-over. i removed the multi-worker from loader, did not show noticeable speed up on my machine. Gonna try on a L4 machine.

File: python/python/lance/torch/data.py (hunk 3 above)
Comment (eddyxu) [reply python/python/lance/torch/data.py]: done

File: python/python/lance/torch/data.py (hunk 2 above)
Comment (eddyxu) [reply python/python/lance/torch/data.py:53]: You can not make a `FixedSizeListArray` from `ChunkedArray`

```
    def test_convert_chunked_fsl(tmp_path: Path):
        arr = pa.chunked_array([[1.0, 2.0, 3.0], [2.0, 3.0, 4.0]], type=pa.float32())
        print(arr)
>       fsl = pa.FixedSizeListArray.from_arrays(arr, 3)

python/tests/torch_tests/test_data.py:266: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   TypeError: Cannot convert pyarrow.lib.ChunkedArray to pyarrow.lib.Array
```

**Review by chebbyChefNEQ — APPROVED** [review]

**Pull request merged**